### PR TITLE
check node.js version before analysis

### LINKED ISSUE #203: Check explicitly for the right version of node being present
As mentioned in #201, before running the typescript analysis it would be nice to have a check that fails with an explicit log in case of node not being present/accessible or not being the right version.

### LINKED ISSUE #339: Analysis halts completely if node bin not found
**I want to report a bug.**
**SonarTS version:** 1.2<br>
**Node.js version:** 6.10<br>
**TypeScript version:** 2.4<br>
**SonarQube version:** 5.6<br>

**Reproducer** 
Run a Gradle analysis on a project with .ts code, where node is not on the PATH and sonar.typescript.node is not set. The Gradle build fails. The analysis is not completed, no report is uploaded.

**Logs**
```
Caused by: java.lang.IllegalStateException: Failed to run external process `node /local/projects/path/to/my/project/sonarts-bundle/node_modules/tslint-sonarts/bin/tsrunner`. Run with -X for more information

Caused by: java.io.IOException: Cannot run program "node": error=2, No such file or directory
```

**Expected behavior**
If node cannot be found (neither available globally nor the bin path set manually), I'd expect for the analysis to gracefully skip the TypeScript analysis. I see the current behaviour an issue in the following scenario:
Some of the projects using my Sonar server might have TypeScript code. If I install the TypeScript plugin right now, analysis will start breaking on these projects, with no changes from their point of view. I need to proactively set the node path on all of them before any projects can start using the plugin.
I'd much prefer if analysis would still continue and succeed, and only log the issue.

What do you think?

P.s. Thanks a lot for fixing #202 !

### PR DESCRIPTION
**Pull request by stas-vilchik**
Fixes #203 
Fixes #339 (no failure of full project analysis if node is not found)

### CODE REVIEW

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable/SonarTSCoreBundle.java [hunk 1]
```diff
@@ -98,6 +97,14 @@ public SonarTSRunnerCommand getSonarTsRunnerCommand(String tsconfigPath, Iterabl
     return runnerCommand;
   }
 
+  /**
+   * Returns path to node.js executable
+   */
+  @Override
+  public String getNodeExecutable() {
+    return configuration.get(TypeScriptPlugin.NODE_EXECUTABLE).get();
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable/SonarTSCoreBundle.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=stas%2Fnode-version&issues=AWCicK81qDrCo5ysgJkn&open=AWCicK81qDrCo5ysgJkn)

[ISSUE_KEY]: AWCicK81qDrCo5ysgJkn

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 2]
```diff
@@ -119,24 +124,60 @@ private void analyze(
...
+      return false;
+    }
+
+    LOG.debug(String.format("Using Node.js %s", version));
+    return true;
+  }
+
+  private @Nullable String getNodeVersion(String nodeExecutable) {
+    try {
+      Process process = Runtime.getRuntime().exec(nodeExecutable + " -v");
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: Vulnerability Vulnerability: Make sure "nodeExecutable" is properly sanitized before use in this OS command. ([squid:S2076](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2076))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=stas%2Fnode-version&issues=AWCicK8kqDrCo5ysgJkl&open=AWCicK8kqDrCo5ysgJkl)

[ISSUE_KEY]: AWCicK8kqDrCo5ysgJkl

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 3]
```diff
@@ -119,24 +124,60 @@ private void analyze(
...
+    } else {
+      LOG.error(String.format("Failed to parse Node.js version, got %s", version));
+      return false;
+    }
+
+    LOG.debug(String.format("Using Node.js %s", version));
+    return true;
+  }
+
+  private @Nullable String getNodeVersion(String nodeExecutable) {
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: Code Smell Code Smell: Make "getNodeVersion" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=stas%2Fnode-version&issues=AWCicK8kqDrCo5ysgJkm&open=AWCicK8kqDrCo5ysgJkm)

[ISSUE_KEY]: AWCicK8kqDrCo5ysgJkm

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 4]
```diff
@@ -119,24 +124,60 @@ private void analyze(
...
+            saveFailures(sensorContext, response.issues, typeScriptRules);
+          } else {
+            LOG.error("Failed to find input file for path `" + response.filepath + "`");
+          }
         }
       }
     }
   }
 
+  private boolean checkNodeVersion(String nodeExecutable) {
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: Code Smell Code Smell: Make "checkNodeVersion" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=stas%2Fnode-version&issues=AWCiiE9r_Q-0dz4JAhyI&open=AWCiiE9r_Q-0dz4JAhyI)

[ISSUE_KEY]: AWCiiE9r_Q-0dz4JAhyI

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java (hunk 2 above)
Comment (SonarTech) [reply sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 5]
```diff
@@ -119,24 +123,60 @@ private void analyze(
...
+    } else {
+      LOG.error(String.format("Failed to parse Node.js version, got %s", version));
+      return false;
+    }
+
+    LOG.debug(String.format("Using Node.js %s", version));
+    return true;
+  }
+
+  private static @Nullable String getNodeVersion(String nodeExecutable) {
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: @stas-vilchik we usually put `@Nullable` on separate line in front of the method

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 6]
```diff
@@ -119,24 +123,60 @@ private void analyze(
...
+    try {
+      Process process = Runtime.getRuntime().exec(nodeExecutable + " -v");
+      return IOUtils.toString(process.getInputStream(), StandardCharsets.UTF_8).trim();
+    } catch (Exception e) {
+      LOG.error("Failed to get Node.js version");
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: @stas-vilchik add one move error message that no analysis will be performed?

File: sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java [hunk 7]
```diff
@@ -222,6 +222,21 @@ public void should_fail_when_failed_external_process() throws Exception {
     executeSensor(sensorContext, testBundle);
   }
 
+  @Test
+  public void should_fail_with_old_node() throws Exception {
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java]: @stas-vilchik update name

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java (hunk 6 above)
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: @stas-vilchik how about moving all logging to one method?

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 8]
```diff
@@ -119,24 +123,60 @@ private void analyze(
...
+            saveFailures(sensorContext, response.issues, typeScriptRules);
+          } else {
+            LOG.error("Failed to find input file for path `" + response.filepath + "`");
+          }
         }
       }
     }
   }
 
+  private static boolean checkNodeVersion(String nodeExecutable) {
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: @stas-vilchik can we have more unit tests for this method?

File: sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java [hunk 9]
```diff
@@ -371,6 +392,11 @@ public ExecutableBundle createAndDeploy(File deployDestination, Configuration co
       public SonarTSRunnerCommand getSonarTsRunnerCommand(String tsconfigPath, Iterable<InputFile> inputFiles, TypeScriptRules typeScriptRules) {
         return new SonarTSRunnerCommand(inputFiles, ruleCheckCommand);
       }
+
+      @Override
+      public String getNodeExecutable() {
+        return customNodeExecutable != null ? customNodeExecutable : node;
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java]: @stas-vilchik so our unit tests will fail if executed on machine with old node?

**Comment by vilchik-elena** [comment]
@stas-vilchik change order of logs:
```
15:43:19.979 DEBUG: Analyzing 1 typescript file(s) with the following configuration file /Users/elena/Projects/demo/ts/proj/tsconfig.json
15:43:19.982 DEBUG: Checking node version
15:43:20.010 DEBUG: Using Node.js v8.9.1
```

**Pull request merged**